File: frontend/src/types/transaction.ts
export type TransactionType = 'INCOME' | 'EXPENSE' | 'TRANSFER';

export type TransactionDto = {
  id: number;
  accountId: number;
  transferAccountId?: number;
  categoryId?: number;
  type: TransactionType;
  amount: number;
  description: string;
  merchant?: string;
  paymentMethod?: string;
  tags: string[];
  transactionDate: string;
  createdAt: string;
  updatedAt: string;
  recurringTransactionId?: number;
};

export type TransactionPage = {
  items: TransactionDto[];
  total: number;
  page: number;
  size: number;
};

export type TransactionListFilter = {
  page?: number;
  size?: number;
  search?: string;
  type?: TransactionType;
  accountId?: number;
  categoryId?: number;
  fromDate?: string;
  toDate?: string;
  minAmount?: number;
  maxAmount?: number;
};

export type CreateTransactionPayload = {
  accountId: number;
  amount: number;
  description: string;
  type: TransactionType;
  categoryId?: number;
  transferAccountId?: number;
  merchant?: string;
  paymentMethod?: string;
  tags?: string[];
  transactionDate?: string;
  recurringTransactionId?: number;
};

export type UpdateTransactionPayload = CreateTransactionPayload;

export type TransactionFormValues = {
  accountId: number;
  amount: number;
  description: string;
  type: TransactionType;
  categoryId?: number;
  transferAccountId?: number;
  merchant?: string;
  paymentMethod?: string;
  tags?: string;
  transactionDate: string;
  recurringTransactionId?: number;
};

File: backend/src/main/java/com/finance/tracker/entity/Transaction.java
package com.finance.tracker.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.OffsetDateTime;

@Entity
@Getter
@Setter
@NoArgsConstructor
@Table(name = "transactions")
public class Transaction {
    public enum TransactionType {
        INCOME,
        EXPENSE,
        TRANSFER
    }

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "account_id", nullable = false)
    private Account account;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "transfer_account_id")
    private Account transferAccount;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @Column(nullable = false)
    private BigDecimal amount;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String description;

    @Column(name = "merchant")
    private String merchant;

    @Column(name = "payment_method")
    private String paymentMethod;

    @Column(name = "tags")
    private String tags;

    @Column(name = "recurring_transaction_id")
    private Long recurringTransactionId;

    @Column(name = "transaction_date", nullable = false)
    private OffsetDateTime transactionDate = OffsetDateTime.now();

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TransactionType type;

    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @PrePersist
    public void prePersist() {
        OffsetDateTime now = OffsetDateTime.now();
        createdAt = now;
        updatedAt = now;
    }

    @PreUpdate
    public void preUpdate() {
        updatedAt = OffsetDateTime.now();
    }
}
